python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -35,16 +46,194 @@ def __init__(self, params, lr):
+                mask = kv_idx_split == local_rank
+                idx = F.boolean_mask(idx, mask)
+            emb_state_dict.update({"ids": idx})
+            emb_state = {}
+            states = (
+                list(self._state[emb.name])
+                if isinstance(self._state[emb.name], tuple)
+                else [self._state[emb.name]]
+            )
+            emb_state = {state.name: state[idx] for state in states}
+            emb_state_dict.update({"states": emb_state})
+            lcoal_state_dict["emb_states"].update({emb.name: emb_state_dict})
+        lcoal_state_dict["params"].update(self._defaults)
+        return lcoal_state_dict
+
+    def load_local_state_dict(self, local_state_dict):
+        """Load the local state from the input state_dict,
+        updating the optimizer as needed..
+
+        Parameters
+        ----------
+        local_state_dict : dict
+            Optimizer state; should be an object returned
+            from a call to local_state_dict()
+
+        See Also
+        --------
+        local_state_dict
+        """
+        for emb_name, emb_state in local_state_dict["emb_states"].items():
+            idx = emb_state["ids"]
+            states = (
+                list(self._state[emb_name])
+                if isinstance(self._state[emb_name], tuple)
+                else [self._state[emb_name]]
+            )
+            if len(emb_state["states"]) != len(states):
+                raise ValueError(
+                    f"loaded state dict has a different number of states"
+                    f" of embedding {emb_name}"
+                )
+            name_to_index = {
+                state.name: index for index, state in enumerate(states, 0)
+            }
+            for name, state in emb_state["states"].items():
+                if name not in name_to_index:
+                    raise ValueError(
+                        "loaded state dict contains a state {name}"
+                        "that can't be found in the optimizer states"
+                    )
+                state_idx = name_to_index[name]
+                state = state.to(states[state_idx].dtype)
+                state = state.to(th.device("cpu"))
+                states[state_idx][idx] = state
+        self._defaults.update(local_state_dict["params"])
+        self.__dict__.update(local_state_dict["params"])
+
+    def save(self, f):
+        """Save the local state_dict to disk on per rank.
+
+        Saved dict contains three parts:
+
+        * 'ids': list of global ids of the nodes/edges selected locally
+        * 'params': hyper parameters of the optimizer
+        * 'states': partial optimizer states indexed by ```ids``` of entire states
+
+        NOTE: This needs to be called on all ranks.
+
+        Parameters
+        ----------
+        f : Union[str, PathLike]
+            The basic path of the file, client id
+            will be attached as full name.
```

Comment:
`client id will be attached as full name` sounds complicated to the users.  Namely, what is "basic path"?  Attached in what way?  Does the "path" actually correspond to multiple files?  All these would bring confusion.
Actually, could we make the behavior simpler?

Reply:
It make sense. I think user don't need to really care about what file name in each machine, they can simply think the file is saved in a ```f``` like locally save. 
To make it clear, use ```Path of the file``` as the description.

Reply:
But it seems that it's still not saving to "the path of the file".  Rather it's saving to some different path, one per trainer.

Could we make it a directory instead?  All the trainer-specific files can go into the directory so it can avoid confusion.

Reply:
Use directory can avoid ambiguity. but the problem is if multiple files are saved under one folder, we can't decide which one to load. And it is hard for  users to understand why they can't save different files under one directory. Because we want user to feel it like a local save similar to PyTorch Save.